src/main.rs

```diff
@@ -0,0 +1,157 @@
+
+    for plugin in config.plugins {
+        let plugin_name = &plugin.name;
+        println!("\n\nNew plugin: {plugin_name}");
+
+        let plugin_add_result = run_plugin_add(&plugin_name);
+
+        match plugin_add_result {
+            // println! macro only takes str primitive literals, not Strings.
+            // So to get around that, we have it expand whatever we pass it, which
+            // happens to be a String.
+            Ok(msg) => println!("{}", msg),
+            Err(msg) => {
+                println!("{}", msg);
+                println!("Skipping installing versions for `{plugin_name}");
+                continue;
+            }
+        };
+
+        // There is likely a way to do this by making the struct mutable, but
+        // I'm not sure what it is and these sets are small enough that cloning
+        // is just easier to keep moving forward.
+        let mut versions_to_install = plugin.versions.clone();
+
+        // We need this to throw an unwrapped default into the hashset of versions
+        // and then to check again later for if we need to actaully set default version.
+        let default_for_later = plugin.default_version.clone();
+
+        if let Some(default_version) = plugin.default_version {
+            versions_to_install.insert(default_version);
+        }
+
+        if versions_to_install.is_empty() {
+            println!("=== No versions configured");
+            continue;
+        } else {
+            for version in versions_to_install {
+                install_version(&plugin_name, &version);
+            }
+        }
+
+        if let Some(default_version) = default_for_later {
+            set_default_version(&plugin_name, &default_version);
+        };
+    }
+}
+
+fn validate_asdf_exists() {
+    match which(ASDF) {
+        Ok(_) => println!("asdf available"),
+        Err(_) => {
+            println!("!!! asdf could not be found in path");
+            std::process::exit(1);
+        }
+    }
+}
+
+// This function returns a result, but! It still panics in some specific cases.
+// Namely, something very unexpected running theh plugin add command, like a segfault,
+// and if the command is terminated by a signal somehow.
+// This is probably a bad practice but for now I do not care.
+fn run_plugin_add(plugin_name: &str) -> Result<String, String> {
+    let output = Command::new(ASDF)
+        .args(["plugin", "add", &plugin_name])
+        .output()
+        .expect("!!! Something went very wrong");
+
+    let status = output.status;
+    let code = status
+        .code()
+        .expect("!!! Unexpectedly terminated by signal");
+
+    match code {
+        0 => Ok(format!("`{plugin_name}` successfully installed")),
+        2 => Ok(format!("`{plugin_name}` is already installed")),
+        1 => Err(format!("`{plugin_name}` could not be installed")),
+        _ => Err(format!("Something went wrong adding `{plugin_name}")),
+    }
+}
```

Comment:
Since you mention error-handling in the comment for this function, here's how it could be changed using anyhow (using eyre would be the same, just s/anyhow/eyre).

```suggestion
fn run_plugin_add(plugin_name: &str) -> anyhow::Result<String> {
    let output = Command::new(ASDF)
        .args(["plugin", "add", &plugin_name])
        .output()?;

    let status = output.status;
    let code = status.code()?;

    match code {
        0 => Ok(format!("`{plugin_name}` successfully installed")),
        2 => Ok(format!("`{plugin_name}` is already installed")),
        1 => Err(anyhow!("`{plugin_name}` could not be installed")),
        _ => Err(anyhow!("Something went wrong adding `{plugin_name}")),
    }
}
```